internal/errors.go

```diff
@@ -2,39 +2,42 @@ package internal
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
 	"io"
 	"strings"
+
+	"github.com/cilium/ebpf/internal/unix"
 )
 
-// ErrorWithLog returns an error which includes logs from the kernel verifier.
+// ErrorWithLog wraps err in a VerifierError that includes the parsed verifier
+// log buffer.
 //
 // The default error output is a summary of the full log. The latter can be
 // accessed via VerifierError.Log or by formatting the error, see Format.
 //
-// A set of heuristics is used to determine whether the log has been truncated.
-func ErrorWithLog(err error, log []byte) *VerifierError {
+// A secondary error (e.g. from a retried load with verifier log enabled)
+// can be provided that is checked for ENOSPC to determine if the kernel
+// truncated the log.
+func ErrorWithLog(err, ve error, log []byte) *VerifierError {
 	const whitespace = "\t\r\v\n "
 
+	// Either error can be ENOSPC depending on whether or now the verifier
+	// was explicitly enabled by the caller.
+	truncated := false
+	if errors.Is(err, unix.ENOSPC) || errors.Is(ve, unix.ENOSPC) {
```

Comment:
What is the rationale to also check err? We've not done this before, and if the log is disabled ENOSPC can't mean that truncation occurred.

Put another way, we should only set truncated if `logEnabled && err == ENOSPC` to be conservative, no?

Nit: could also be `truncate := errors.Is ...`.

Reply:
If the caller sets `LogLevel` > 0, `ve` is never assigned to. (the second load will be skipped) If the buffer is too small, ENOSPC will be in the first `err`.

The only other ENOSPC case I can find in PROG_LOAD is when a btf ID fails to be allocated for the prog (id's 1..INT_MAX exhausted). If a machine gets into this state, a false-positive truncated flag sounds minor in comparison.

That's not to say I don't agree. Thinking about it, setting the truncated flag doesn't belong in `ErrorWithLog` to begin with. Only setting Truncated when we've requested a log does sound safer.